Ensure that intermediate directories exist when unpacking an entry

ranma42 opened the pull request:
The `unpack` function assumes that the directory in which the file is
being extracted exists, while most `tar` tools will automatically
create the intermediate directories if they are missing.

This would have avoided #1092.

brson commented:
@bors r+

Note though that we still cannot remove the directory entries from the tarballs produced by Rust, because they need to be compatible with older rustups. I'm not sure how long a suitable deprecation period is.

bors commented:
:pushpin: Commit 5ae90c4 has been approved by `brson`

<!-- @bors r=brson 5ae90c4b52e81b18c560acdfa70565fff37465bf -->

bors commented:
:hourglass: Testing commit 5ae90c4 with merge 4f98cc0...

ranma42 commented:
Yes, I have no plans for removing them and the Rust port of rust-installer will keep them as well for the time being. Removing them was just convenient when sorting the paths.

bors commented:
:broken_heart: Test failed - [status-appveyor](https://ci.appveyor.com/project/brson/rustup-rs/build/1.0.1068)

brson commented:
Legit errors.

ranma42 committed:
Create full path to file entries

The `unpack` function assumes that the directory in which the file is
being extracted exists, while most `tar` tools will automatically
create the intermediate directories if they are missing.

This would have avoided #1092.

ranma42 commented:
It is easier to fix without updating tar. Updated title and first post to get reasonable commit messages upon merging.

brson commented:
@bors r+

I can imagine this repeated syscall wanting to be optimized someday, but today I'm not too worried. rustup generally needs to be optimized.

bors commented:
:pushpin: Commit ab40997 has been approved by `brson`

<!-- @bors r=brson ab40997834b902358e7290db474b6b1d01e07214 -->

bors commented:
:hourglass: Testing commit ab40997 with merge 6a5f6f2...

bors commented:
:broken_heart: Test failed - [status-appveyor](https://ci.appveyor.com/project/brson/rustup-rs/build/1.0.1091)

ranma42 commented:
Is the failure spurious? The non-merge AppVeyor build succeeded

Diggsey commented:
Yes - @bors r=brson retry

bors commented:
:bulb: This pull request was already approved, no need to approve it again.

- There's another pull request that is currently being tested, blocking this pull request: #1086

bors commented:
:pushpin: Commit ab40997 has been approved by `brson`

<!-- @bors r=brson ab40997834b902358e7290db474b6b1d01e07214 -->

bors commented:
:hourglass: Testing commit ab40997 with merge 3605b7e...

bors commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/brson/rustup-rs/build/1.0.1116), [status-travis](https://travis-ci.org/rust-lang-nursery/rustup.rs/builds/233398183?utm_source=github_status&utm_medium=notification)
Approved by: brson
Pushing 3605b7e8af02bde78c5267c33da2f39591e198b1 to master...
